### PR TITLE
[NOT FOR MERGE] Proposal: Add some infra to tinygrad for modern LLM pretraining

### PR DESCRIPTION
**Pull request by luyanaa**
> Disclaimer: I have no adequate servers so these patches didn't go through anything other than smoke tests. Also, they're immature so they couldn't be treated as a formal pull request. 

Seems that tinygrad is actively working on some multi-GPU or multi-node training infra. I would like to propose some features that seems to be good for it?

1. tinygrad needs to support AMD and NVIDIA GPUs, while not fully utilizing CUDA or ROCm, so RCCL and NCCL is limited to some extent? (Please tell me if I'm wrong, but https://docs.pytorch.org/docs/main/distributed.html). Many GPU clusters are managed with slurm or PBS (say, the one I've used in The University of Tokyo, the miyabi-G), and MPI provides a good bootstrap for spawning multi-nodes. 
2. Also, if RDMA is not available (for some strange beowulf clusters I have built or designed..., say, clusters with 10G NiC, a project to utilize vacant 2080Ti x40 somewhere)
3. Seems that currently there's sharding in models (not sure if it could be taken as pipeline parallel), but I think maybe a light wrapper for tensor parallel is also good to have?
4. DeepSpeed Zero-1/2/3 and CPU offload is definitely good to have. 
5. A light wrapper on gradient checkpointing is good to have too. 

Sorry for the sudden "not that PR", more of a request, not sure if I have opportunities to turn them into more reasonable patches. 

Thanks.

### CODE REVIEW

**Comment by geohot** [comment]
do not use ai
